pkg/provider/cloud/openstack/provider.go

```diff
@@ -622,8 +622,15 @@ func linkSubnetToRouter(netClient *gophercloud.ServiceClient, cluster *kubermati
 		if err != nil {
 			return err
 		}
-		kubernetes.AddFinalizer(cluster, finalizer)
+		// Add the Link finalizer
+		_, err = update(ctx, cluster.Name, func(cluster *kubermaticv1.Cluster) {
+			kubernetes.AddFinalizer(cluster, finalizer)
+		})
+		if err != nil {
+			return fmt.Errorf("failed to add RouterSubnetLinkCleanupFinalizer in the cluster spec: %w", err)
```

Comment:
Replace the concrete finalizer with %s and use `finalizer` to make the error correct.